File: operator/attestor/event_listener.go
package attestor

import (
	"fmt"

	"github.com/prometheus/client_golang/prometheus"
)

type EventListener interface {
	OnMissedMQBlock(rollupId uint32)
	OnBlockMismatch(rollupId uint32)
	OnBlockReceived(rollupId uint32)
	ObserveLastBlockReceived(rollupId uint32, blockNumber uint64)
	ObserveLastBlockReceivedTimestamp(rollupId uint32, timestamp uint64)
	ObserveInitializationInitialBlockNumber(rollupId uint32, blockNumber uint64)
}

const OperatorNamespace = "sffl_operator"
const AttestorSubsystem = "attestor"

type SelectiveEventListener struct {
	OnMissedMQBlockCb                         func(rollupId uint32)
	OnBlockMismatchCb                         func(rollupId uint32)
	OnBlockReceivedCb                         func(rollupId uint32)
	ObserveLastBlockReceivedCb                func(rollupId uint32, blockNumber uint64)
	ObserveLastBlockReceivedTimestampCb       func(rollupId uint32, timestamp uint64)
	ObserveInitializationInitialBlockNumberCb func(rollupId uint32, blockNumber uint64)
}

func (l *SelectiveEventListener) OnMissedMQBlock(rollupId uint32) {
	if l.OnMissedMQBlockCb != nil {
		l.OnMissedMQBlockCb(rollupId)
	}
}

func (l *SelectiveEventListener) OnBlockMismatch(rollupId uint32) {
	if l.OnBlockMismatchCb != nil {
		l.OnBlockMismatchCb(rollupId)
	}
}

func (l *SelectiveEventListener) OnBlockReceived(rollupId uint32) {
	if l.OnBlockReceivedCb != nil {
		l.OnBlockReceivedCb(rollupId)
	}
}

func (l *SelectiveEventListener) ObserveLastBlockReceived(rollupId uint32, blockNumber uint64) {
	if l.ObserveLastBlockReceivedCb != nil {
		l.ObserveLastBlockReceivedCb(rollupId, blockNumber)
	}
}

func (l *SelectiveEventListener) ObserveLastBlockReceivedTimestamp(rollupId uint32, timestamp uint64) {
	if l.ObserveLastBlockReceivedTimestampCb != nil {
		l.ObserveLastBlockReceivedTimestampCb(rollupId, timestamp)
	}
}

func (l *SelectiveEventListener) ObserveInitializationInitialBlockNumber(rollupId uint32, blockNumber uint64) {
	if l.ObserveInitializationInitialBlockNumberCb != nil {
		l.ObserveInitializationInitialBlockNumberCb(rollupId, blockNumber)
	}
}

func MakeAttestorMetrics(registry *prometheus.Registry) (EventListener, error) {
	numMissedMqBlocks := prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Namespace: OperatorNamespace,
			Subsystem: AttestorSubsystem,
			Name:      "num_of_missed_mq_blocks",
			Help:      "The number of late blocks from MQ",
		},
		[]string{"rollup_id"},
	)

	if err := registry.Register(numMissedMqBlocks); err != nil {
		return nil, fmt.Errorf("error registering numMissedMqBlocks counter: %w", err)
	}

	numBlocksMismatched := prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Namespace: OperatorNamespace,
			Subsystem: AttestorSubsystem,
			Name:      "num_of_mismatched_blocks",
			Help:      "The number of blocks from MQ mismatched with RPC ones.",
		},
		[]string{"rollup_id"},
	)

	if err := registry.Register(numBlocksMismatched); err != nil {
		return nil, fmt.Errorf("error registering numBlocksMismatched counter: %w", err)
	}

	numBlocksReceived := prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Namespace: OperatorNamespace,
			Subsystem: AttestorSubsystem,
			Name:      "num_of_received_blocks",
			Help:      "The number of blocks received from RPC.",
		},
		[]string{"rollup_id"},
	)

	if err := registry.Register(numBlocksReceived); err != nil {
		return nil, fmt.Errorf("error registering numBlocksReceived counter: %w", err)
	}

	lastBlockReceived := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Namespace: OperatorNamespace,
			Name:      "last_block_received",
			Help:      "Last block received per rollup ID",
		},
		[]string{"rollup_id"},
	)

	if err := registry.Register(lastBlockReceived); err != nil {
		return nil, fmt.Errorf("error registering lastBlockReceived gauge: %w", err)
	}

	lastBlockReceivedTimestamp := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Namespace: OperatorNamespace,
			Name:      "last_block_received_timestamp",
			Help:      "Timestamp of last block received per rollup ID",
		},
		[]string{"rollup_id"},
	)

	if err := registry.Register(lastBlockReceivedTimestamp); err != nil {
		return nil, fmt.Errorf("error registering lastBlockReceivedTimestamp gauge: %w", err)
	}

	initializationInitialBlockNumber := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Namespace: OperatorNamespace,
			Name:      "initialization_initial_block_number",
			Help:      "Initialization initial block number per rollup ID",
		},
		[]string{"rollup_id"},
	)

	if err := registry.Register(initializationInitialBlockNumber); err != nil {
		return nil, fmt.Errorf("error registering initializationInitialBlockNumber gauge: %w", err)
	}

	return &SelectiveEventListener{
		OnMissedMQBlockCb: func(rollupId uint32) {
			numMissedMqBlocks.WithLabelValues(fmt.Sprint(rollupId)).Inc()
		},
		OnBlockMismatchCb: func(rollupId uint32) {
			numBlocksMismatched.WithLabelValues(fmt.Sprint(rollupId)).Inc()
		},
		ObserveLastBlockReceivedCb: func(rollupId uint32, blockNumber uint64) {
			lastBlockReceived.WithLabelValues(fmt.Sprint(rollupId)).Set(float64(blockNumber))
		},
		ObserveLastBlockReceivedTimestampCb: func(rollupId uint32, timestamp uint64) {
			lastBlockReceivedTimestamp.WithLabelValues(fmt.Sprint(rollupId)).Set(float64(timestamp))
		},
		ObserveInitializationInitialBlockNumberCb: func(rollupId uint32, blockNumber uint64) {
			initializationInitialBlockNumber.WithLabelValues(fmt.Sprint(rollupId)).Set(float64(blockNumber))
		},
	}, nil
}

File: operator-rs/src/attestor/event_listener.rs
use prometheus::{Registry, Counter, CounterVec, Gauge, GaugeVec, opts};
use anyhow::Result;

pub trait EventListener: Send + Sync {
    fn on_missed_mq_block(&self, rollup_id: u32);
    fn on_block_mismatch(&self, rollup_id: u32);
    fn on_block_received(&self, rollup_id: u32);
    fn observe_last_block_received(&self, rollup_id: u32, block_number: u64);
    fn observe_last_block_received_timestamp(&self, rollup_id: u32, timestamp: u64);
    fn observe_initialization_initial_block_number(&self, rollup_id: u32, block_number: u64);
}

pub struct SelectiveEventListener {
    on_missed_mq_block_cb: Option<Box<dyn Fn(u32) + Send + Sync>>,
    on_block_mismatch_cb: Option<Box<dyn Fn(u32) + Send + Sync>>,
    on_block_received_cb: Option<Box<dyn Fn(u32) + Send + Sync>>,
    observe_last_block_received_cb: Option<Box<dyn Fn(u32, u64) + Send + Sync>>,
    observe_last_block_received_timestamp_cb: Option<Box<dyn Fn(u32, u64) + Send + Sync>>,
    observe_initialization_initial_block_number_cb: Option<Box<dyn Fn(u32, u64) + Send + Sync>>,
}

impl Default for SelectiveEventListener {
    fn default() -> Self {
        Self {
            on_missed_mq_block_cb: None,
            on_block_mismatch_cb: None,
            on_block_received_cb: None,
            observe_last_block_received_cb: None,
            observe_last_block_received_timestamp_cb: None,
            observe_initialization_initial_block_number_cb: None,
        }
    }
}

impl EventListener for SelectiveEventListener {
    fn on_missed_mq_block(&self, rollup_id: u32) {
        if let Some(cb) = &self.on_missed_mq_block_cb {
            cb(rollup_id);
        }
    }

    fn on_block_mismatch(&self, rollup_id: u32) {
        if let Some(cb) = &self.on_block_mismatch_cb {
            cb(rollup_id);
        }
    }

    fn on_block_received(&self, rollup_id: u32) {
        if let Some(cb) = &self.on_block_received_cb {
            cb(rollup_id);
        }
    }

    fn observe_last_block_received(&self, rollup_id: u32, block_number: u64) {
        if let Some(cb) = &self.observe_last_block_received_cb {
            cb(rollup_id, block_number);
        }
    }

    fn observe_last_block_received_timestamp(&self, rollup_id: u32, timestamp: u64) {
        if let Some(cb) = &self.observe_last_block_received_timestamp_cb {
            cb(rollup_id, timestamp);
        }
    }

    fn observe_initialization_initial_block_number(&self, rollup_id: u32, block_number: u64) {
        if let Some(cb) = &self.observe_initialization_initial_block_number_cb {
            cb(rollup_id, block_number);
        }
    }
}

const OPERATOR_NAMESPACE: &str = "sffl_operator";
const ATTESTOR_SUBSYSTEM: &str = "attestor";

pub fn make_attestor_metrics(registry: &Registry) -> Result<impl EventListener> {
    let num_missed_mq_blocks = CounterVec::new(
        opts!("num_of_missed_mq_blocks", "The number of late blocks from MQ")
            .namespace(OPERATOR_NAMESPACE)
            .subsystem(ATTESTOR_SUBSYSTEM),
        &["rollup_id"],
    )?;
    registry.register(Box::new(num_missed_mq_blocks.clone()))?;

    let num_blocks_mismatched = CounterVec::new(
        opts!("num_of_mismatched_blocks", "The number of blocks from MQ mismatched with RPC ones")
            .namespace(OPERATOR_NAMESPACE)
            .subsystem(ATTESTOR_SUBSYSTEM),
        &["rollup_id"],
    )?;
    registry.register(Box::new(num_blocks_mismatched.clone()))?;

    let num_blocks_received = CounterVec::new(
        opts!("num_of_received_blocks", "The number of blocks received from RPC")
            .namespace(OPERATOR_NAMESPACE)
            .subsystem(ATTESTOR_SUBSYSTEM),
        &["rollup_id"],
    )?;
    registry.register(Box::new(num_blocks_received.clone()))?;

    let last_block_received = GaugeVec::new(
        opts!("last_block_received", "Last block received per rollup ID")
            .namespace(OPERATOR_NAMESPACE),
        &["rollup_id"],
    )?;
    registry.register(Box::new(last_block_received.clone()))?;

    let last_block_received_timestamp = GaugeVec::new(
        opts!("last_block_received_timestamp", "Timestamp of last block received per rollup ID")
            .namespace(OPERATOR_NAMESPACE),
        &["rollup_id"],
    )?;
    registry.register(Box::new(last_block_received_timestamp.clone()))?;

    let initialization_initial_block_number = GaugeVec::new(
        opts!("initialization_initial_block_number", "Initialization initial block number per rollup ID")
            .namespace(OPERATOR_NAMESPACE),
        &["rollup_id"],
    )?;
    registry.register(Box::new(initialization_initial_block_number.clone()))?;

    Ok(SelectiveEventListener {
        on_missed_mq_block_cb: Some(Box::new(move |rollup_id| {
            num_missed_mq_blocks.with_label_values(&[&rollup_id.to_string()]).inc();
        })),
        on_block_mismatch_cb: Some(Box::new(move |rollup_id| {
            num_blocks_mismatched.with_label_values(&[&rollup_id.to_string()]).inc();
        })),
        on_block_received_cb: Some(Box::new(move |rollup_id| {
            num_blocks_received.with_label_values(&[&rollup_id.to_string()]).inc();
        })),
        observe_last_block_received_cb: Some(Box::new(move |rollup_id, block_number| {
            last_block_received.with_label_values(&[&rollup_id.to_string()]).set(block_number as f64);
        })),
        observe_last_block_received_timestamp_cb: Some(Box::new(move |rollup_id, timestamp| {
            last_block_received_timestamp.with_label_values(&[&rollup_id.to_string()]).set(timestamp as f64);
        })),
        observe_initialization_initial_block_number_cb: Some(Box::new(move |rollup_id, block_number| {
            initialization_initial_block_number.with_label_values(&[&rollup_id.to_string()]).set(block_number as f64);
        })),
    })
}